### PR TITLE
Adding game german_whist_foregame

### PR DESCRIPTION
**Pull request by willmcgowan**
German Whist is a 2 player imperfect information trick-taking card game.  The rules are outlined here https://en.wikipedia.org/wiki/German_whist#:~:text=German%20whist%20is%20a%20variant,is%20probably%20of%20British%20origin.

The foregames are scored by accessing a tablebase that stores the game theoretic value of all possible endgames. The code to generate this tablebase is the german_whist_endgame executable. On my device it takes around 2 hours to generate on 4 cores. It requires BMI2 intrinsics to run. The tablebase will take up around 1.8GB of ram when in use, and the code generating the tablebase will take up slightly more than that while running. Once the tablebase is built modify the kTTablePath variable to its path. 
There is an executable in the examples file that allows play against an ISMCTS agent on your device.

### CODE REVIEW

**Comment by lanctot** [comment]
Thanks @willmcgowan ! Can you please sign the CLA?

File: open_spiel/CMakeLists.txt [hunk 1]
```diff
@@ -34,7 +34,7 @@ set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
 # Set default build type.
 set (BUILD_TYPE $ENV{BUILD_TYPE})
 if(NOT BUILD_TYPE)
-   set(BUILD_TYPE Testing
+   set(BUILD_TYPE Release
```
Comment (lanctot) [inline open_spiel/CMakeLists.txt:37]: Can you set this back to Testing?

File: open_spiel/CMakeLists.txt [hunk 2]
```diff
@@ -50,14 +50,14 @@ if(${BUILD_TYPE} STREQUAL "Testing")
   # A build used for running tests: keep all runtime checks (assert,
   # SPIEL_CHECK_*, SPIEL_DCHECK_*), but turn on some speed optimizations,
   # otherwise tests run for too long.
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2 -march=x86-64-v3")
```
Comment (lanctot) [inline open_spiel/CMakeLists.txt:53]: Can you revert this?

File: open_spiel/CMakeLists.txt [hunk 3]
```diff
@@ -50,14 +50,14 @@ if(${BUILD_TYPE} STREQUAL "Testing")
...
 endif()
 
 if(${BUILD_TYPE} STREQUAL "Release")
   # Optimized release build: turn off debug runtime checks (assert,
   # SPIEL_DCHECK_*) and turn on highest speed optimizations.
   # The difference in perfomance can be up to 10x higher.
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DNDEBUG -O3")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DNDEBUG -O3 -march=x86-64-v3")
```
Comment (lanctot) [inline open_spiel/CMakeLists.txt:60]: And revert this too

File: open_spiel/games/german_whist_foregame/german_whist_foregame.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
+#ifndef OPEN_SPIEL_GAMES_GERMAN_WHIST_FOREGAME_H
+#define OPEN_SPIEL_GAMES_GERMAN_WHIST_FOREGAME_H
+
+#include <array>
+#include <memory>
+#include <string>
+#include <vector>
+#include <random>
+#include <x86intrin.h>
```
Comment (lanctot) [inline open_spiel/games/german_whist_foregame/german_whist_foregame.h:9]: Is this a standard header?

Is it used anywhere in this file? Code submitted to the master repos should be portable as we also support MacOS run on different architectures.

File: open_spiel/games/german_whist_foregame/german_whist_foregame.h [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+#include "open_spiel/policy.h"
+#include "open_spiel/spiel.h"
+#include "open_spiel/spiel_utils.h"
+
+//The imperfect information part of 2 player whist variant
+//https://en.wikipedia.org/wiki/German_Whist
+//
+//
+
+//
```
Comment (lanctot) [inline open_spiel/games/german_whist_foregame/german_whist_foregame.h:24]: Remove

File: open_spiel/games/german_whist_foregame/german_whist_foregame.h (hunk 4 above)
Comment (lanctot) [reply open_spiel/games/german_whist_foregame/german_whist_foregame.h:9]: If you are using x86 architecture optimizations, what I suggest is that you keep this file as it is now (i.e. for your research) and submit a different one that is completely general and can run on non-x86 architectures.

**Review by lanctot — CHANGES_REQUESTED** [review]
I left a few quick comments for now but have not looked too closely yet.

**Comment by willmcgowan** [comment]
Ok I will need to reimplement portions of the code. I will close this for now and submit a general implementation

**Pull request closed without merging**